cpp/oneapi/dal/algo/kmeans/test/spmd_backend_template.hpp

```diff
@@ -0,0 +1,295 @@
+                     "[spmd][smoke]",
+                     kmeans_types) {
+    // SPMD mode is not implemented for CPU. The following `SKIP_IF` should be
+    // removed once it's supported for CPU. The same for the rest of tests cases.
+    SKIP_IF(this->get_policy().is_cpu());
+    SKIP_IF(this->not_float64_friendly());
+
+    this->check_if_results_same_on_all_ranks();
+}
+TEMPLATE_LIST_TEST_M(kmeans_spmd_backend_test,
+                     "distributed kmeans empty clusters test",
+                     "[spmd][smoke]",
+                     kmeans_types) {
+    SKIP_IF(this->get_policy().is_cpu());
+    SKIP_IF(this->not_float64_friendly());
+
+    this->check_empty_clusters();
+}
+
+TEMPLATE_LIST_TEST_M(kmeans_spmd_backend_test,
+                     "distributed kmeans smoke train/infer test",
+                     "[spmd][smoke]",
+                     kmeans_types) {
+    SKIP_IF(this->get_policy().is_cpu());
+    SKIP_IF(this->not_float64_friendly());
+
+    this->check_on_smoke_data();
+}
+
+TEMPLATE_LIST_TEST_M(kmeans_spmd_backend_test,
+                     "distributed kmeans train/infer on gold data",
+                     "[spmd][smoke]",
+                     kmeans_types) {
+    SKIP_IF(this->get_policy().is_cpu());
+    SKIP_IF(this->not_float64_friendly());
+
+    this->check_on_gold_data();
+}
+
+TEMPLATE_LIST_TEST_M(kmeans_spmd_backend_test,
+                     "distributed kmeans block test",
+                     "[spmd][block][nightly]",
+                     kmeans_types) {
+    SKIP_IF(this->get_policy().is_cpu());
+    SKIP_IF(this->not_float64_friendly());
+
+    this->check_on_large_data_with_one_cluster();
+}
+
+TEMPLATE_LIST_TEST_M(kmeans_spmd_backend_test,
+                     "distributed higgs: samples=1M, iters=3",
+                     "[kmeans][spmd][higgs][external-dataset]",
+                     kmeans_types) {
+    SKIP_IF(this->get_policy().is_cpu());
+    SKIP_IF(this->not_float64_friendly());
+
+    const std::int64_t iters = 3;
+    const std::string higgs_path = "workloads/higgs/dataset/higgs_1m_test.csv";
+
+    SECTION("clusters=10") {
+        this->test_on_dataset(higgs_path, 10, iters, 3.1997724684, 14717484.0);
+    }
+
+    SECTION("clusters=100") {
+        this->test_on_dataset(higgs_path, 100, iters, 2.7450205195, 10704352.0);
+    }
+
+    SECTION("cluster=250") {
+        this->test_on_dataset(higgs_path, 250, iters, 2.5923397174, 9335216.0);
+    }
+}
+
+TEMPLATE_LIST_TEST_M(kmeans_spmd_backend_test,
+                     "distributed susy: samples=0.5M, iters=10",
+                     "[kmeans][nightly][spmd][susy][external-dataset]",
+                     kmeans_types) {
+    SKIP_IF(this->get_policy().is_cpu());
+    SKIP_IF(this->not_float64_friendly());
+
+    const std::int64_t iters = 10;
+    const std::string susy_path = "workloads/susy/dataset/susy_test.csv";
+
+    SECTION("clusters=10") {
+        this->test_on_dataset(susy_path, 10, iters, 1.7730860782, 3183696.0);
```

Comment:
Have such numbers been validated with scikit?

Reply:
It was validated for CPU/GPU/SPMD consistency